src/commands/build-util.ts

```diff
@@ -1,19 +1,135 @@
+        input: output[destPath].code,
+        cwd: process.cwd(),
+      });
+      if (stderr && messageBus) {
+        messageBus.emit('WORKER_MSG', {id, level: 'warn', msg: `${srcPath}\n${stderr}`});
+      }
+      output[destPath].code = stdout;
+    } catch (err) {
+      if (messageBus) {
+        messageBus.emit('WORKER_MSG', {id, level: 'error', msg: `${srcPath}\n${err.stderr}`});
+        messageBus.emit('WORKER_UPDATE', {id, state: ['ERROR', 'red']});
+      }
+    }
+  }
+
+  // 2. transform with Snowpack plugins (main build pipeline)
+  for (const step of config.__buildPipeline[srcExt.expandedExt] ||
+    config.__buildPipeline[srcExt.baseExt] ||
+    []) {
+    // TODO: remove transform() from plugin API when no longer used
+    if (step.transform) {
+      const urlPath = destPath.substr(destPath.length + 1);
+      const {result} = await step.transform({
+        contents: output[destPath].code,
+        urlPath,
+        isDev: false,
+      });
+      output[destPath].code = result;
+    }
+
+    if (step.build) {
+      let result: BuildResult;
+      try {
+        result = await step.build({
+          code: output[destPath].code,
+          contents: output[destPath].code,
+          filePath: srcPath,
+          isDev: false,
+        });
+      } catch (err) {
+        if (messageBus) {
+          messageBus.emit('WORKER_MSG', {id: step.name, level: 'error', msg: err.message});
+          messageBus.emit('WORKER_UPDATE', {id: step.name, state: ['ERROR', 'red']});
+        }
+        process.exit(1);
+      }
+
+      if (typeof result === 'string') {
+        // Path A: single-output (assume extension is same)
+        output[destPath].code = result;
+      } else if (typeof result === 'object' && !result.result) {
+        // Path B: multi-file output ({ js: [string], css: [string], … })
+        Object.entries(result as {[ext: string]: string}).forEach(([ext, code]) => {
+          const newDest = replaceExt(destPath, ext);
+          output[newDest] = {baseExt: ext, expandedExt: ext, code, locOnDisk: srcPath};
+        });
+      } else if (typeof result === 'object' && result.result) {
+        // Path C: DEPRECATED output ({ result, resources })
+        output[destPath].code = result.result;
+
+        // handle CSS output for Svelte/Vue
+        if (typeof result.resources === 'object' && result.resources.css) {
+          const cssFile = replaceExt(destPath, '.css');
+          output[cssFile] = {
+            baseExt: '.css',
+            expandedExt: '.css',
+            code: result.resources.css,
+            locOnDisk: srcPath,
+          };
+        }
+      }
+
+      // filter out unused extensions (i.e. don’t emit source files to build)
+      const outputs = Array.isArray(step.output) ? step.output : [step.output];
+      Object.keys(output).forEach((key) => {
+        const {baseExt} = getExt(key);
+        if (!outputs.includes(baseExt)) {
+          delete output[key];
+        }
+      });
+    }
```

Comment:
Another focus: prevent source files from leaking out during the build. We remove them by scanning `output`.